llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeFlashAttentionPass.cpp

```diff
@@ -0,0 +1,420 @@
+  Value keySlice = builder.create<tensor::ExtractSliceOp>(
+      loc, tensorType, key, offsets, sizes, strides);
+  Value valueSlice = builder.create<tensor::ExtractSliceOp>(
+      loc, tensorType, value, offsets, sizes, strides);
+
+  offsets = SmallVector<OpFoldResult>(queryShape.size(), zero);
+  offsets[0] = ivs[0];
+  Value querySlice = builder.create<tensor::ExtractSliceOp>(
+      loc, tensorType, query, offsets, sizes, strides);
+  Value outputSlice = builder.create<tensor::ExtractSliceOp>(
+      loc, tensorType, output, offsets, sizes, strides);
+
+  return std::make_tuple(keySlice, valueSlice, querySlice, outputSlice);
+}
+
+static std::tuple<Value, Value, Value>
+insertSlices(Value newResult, Value result, Value newMax, Value max,
+             Value newSum, Value sum, ArrayRef<int64_t> queryShape,
+             ArrayRef<Value> ivs, Value sequenceTileLength, Location loc,
+             OpBuilder &builder) {
+  auto one = builder.getIndexAttr(1);
+  auto zero = builder.getIndexAttr(0);
+  auto headDimension = builder.getIndexAttr(queryShape.back());
+  SmallVector<OpFoldResult> strides(queryShape.size(), one);
+  SmallVector<OpFoldResult> sizes(queryShape.size(), one);
+  SmallVector<OpFoldResult> offsets(queryShape.size(), zero);
+  sizes[1] = sequenceTileLength;
+  sizes[2] = headDimension;
+  offsets[0] = ivs[0];
+  Value updatedAcc = builder.create<tensor::InsertSliceOp>(
+      loc, newResult, result, offsets, sizes, strides);
+  offsets = SmallVector<OpFoldResult>{zero};
+  sizes = SmallVector<OpFoldResult>{sequenceTileLength};
+  strides = SmallVector<OpFoldResult>{one};
+  Value updatedMax = builder.create<tensor::InsertSliceOp>(
+      loc, newMax, max, offsets, sizes, strides);
+  Value updatedSum = builder.create<tensor::InsertSliceOp>(
+      loc, newSum, sum, offsets, sizes, strides);
+  return std::make_tuple(updatedAcc, updatedMax, updatedSum);
+}
+
+static scf::LoopNest createLoopNest(SmallVectorImpl<Value> &ivs, Value lb,
+                                    Value step, Value ub, ValueRange args,
+                                    Location loc, OpBuilder &builder) {
+  SmallVector<Value> lbs{lb};
+  SmallVector<Value> steps{step};
+  SmallVector<Value> ubs{ub};
+  scf::LoopNest loopNest = scf::buildLoopNest(
+      builder, loc, lbs, ubs, steps, args,
+      [&](OpBuilder &nestedBuilder, Location loc, ValueRange outputIvs,
+          ValueRange iterArgs) -> scf::ValueVector { return iterArgs; });
+  for (scf::ForOp loop : loopNest.loops) {
+    ivs.push_back(loop.getInductionVar());
+  }
+  return loopNest;
+}
+
+class ReifyFlashAttentionFwdTransform final
+    : public OpRewritePattern<FlashAttentionFwdOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(FlashAttentionFwdOp attnOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = attnOp.getLoc();
+    rewriter.setInsertionPoint(attnOp);
```

Comment:
It is highly desirable to use `OpBuilder::InsertionGuard` before modifying the insertion point of the rewriter passed by reference from above, the caller may not expect modifications to the insertion point.